src/commands/notify.rs

```diff
@@ -0,0 +1,287 @@
+                    Ttl::ZERO,
+                ),
+            );
+            msg.push(soa)
+                .map_err(|e| format!("could not add SOA record: {e}"))?;
+        }
+
+        let msg = msg.additional();
+
+        let tsig = self
+            .tsig
+            .as_ref()
+            .map(|tsig| {
+                Key::new(tsig.algorithm, &tsig.key, tsig.name.clone(), None, None)
+                    .map_err(|e| format!("TSIG key is invalid: {e}"))
+            })
+            .transpose()?;
+
+        let msg = msg.into_message();
+
+        println!("# Sending packet:");
+        println!("{}", msg.display_dig_style());
+
+        if self.debug {
+            println!("Hexdump of notify packet:");
+            println!("{}", base16::encode_display(&msg));
+        }
+
+        let resolver = StubResolver::new();
+
+        for server in &self.servers {
+            println!("# sending to {}", server);
+
+            let Ok(name) = Name::<Vec<u8>>::from_str(server) else {
+                eprintln!("warning: invalid domain name \"{server}\", skipping.");
+                continue;
+            };
+
+            let Ok(hosts) = resolver.lookup_host(&name).await else {
+                eprintln!("warning: could not resolve host \"{name}\", skipping.");
+                continue;
+            };
+
+            if hosts.is_empty() {
+                eprintln!("skipping bad address: {name}: Name or service not known");
+                continue;
+            }
+
+            for socket in hosts.port_iter(self.port) {
+                let resp = match &tsig {
+                    Some(tsig) => {
+                        self.notify_host_with_tsig(socket, msg.clone(), tsig.clone())
+                            .await
+                    }
+                    None => self.notify_host_without_tsig(socket, msg.clone()).await,
+                };
+
+                match resp {
+                    Ok(resp) => {
+                        println!("# reply from {server} at {socket}:");
+                        println!("{resp}");
+                    },
+                    Err(e) => {
+                        eprintln!("{e}");
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn notify_host_with_tsig(
+        &self,
+        socket: SocketAddr,
+        msg: Message<Vec<u8>>,
+        key: Key,
+    ) -> Result<Response<Bytes>, Error> {
+        let mut config = dgram::Config::new();
+        config.set_max_retries(self.retries as u8);
+        let connection = dgram::Connection::with_config(UdpConnect::new(socket), config);
+        let connection = tsig::Connection::new(key, connection);
+
+        let req = RequestMessage::new(msg).unwrap();
+        let mut req = connection.send_request(req);
+    
+        let when = Local::now();
+        let msg = req.get_response().await.map_err(|e| format!("warning: could not send message to {socket}: {e}"))?;
```

Comment:
I would rephrase this error, as I think it's misleading. The request was sent (at least in my test), but the error comes from validating the response from the server. So maybe more like: `warning: erroneous response from {socket}: {e}`. I came across this when using tsig, but the server didn't use tsig in the response.

Reply:
Or will there be different errors? Some from failed sending and some from failed receiving?

Reply:
Both are possible yeah and unfortunately not clearly distinguished. I'll update the text to mention both possibilities.